### PR TITLE
feat: add data quality warning for serving size without digit

### PR DESCRIPTION
**Pull request by benbenben2**
Tested on Gitpod

Related to:
https://github.com/openfoodfacts/openfoodfacts-server/issues/5163
https://github.com/openfoodfacts/openfoodfacts-server/issues/5441
https://github.com/openfoodfacts/openfoodfacts-server/issues/6473

**Questions (and answers):**
1. should it involve nutrition_data_prepared as well?
    -> **decision:** yes
2. raise quality error tag - only when nutrition_data_per or nutrition_data_prepared_per are "on" (option 1) - vs - if string without digit value is in serving_size (even if nutrition_data_per or nutrition_data_prepared_per are missing) (option 2)
    -> **decision:** all (option 2) because as stated by @CharlesNepote  "when users are often seeing this value they can think it's a good value" (https://github.com/openfoodfacts/openfoodfacts-server/issues/5163)
3. how come serving_quantity=0 when it is "serving" or any string?
    -> in `lib/ProductOpener/Food.pm`, the subroutine `normalize_serving_size` extracts any digit and return it or 0 if nothing is found. 
3.1. can this be used to setup the error?
   -> it cannot be serving_quantity (using normalize_serving_size) because serving_quantity is undefined if the value is something like "serving 1 and serving 2 and serving 3", which contain digit
4. what name to give to the error?
    -> existing one is "en:serving-size-in-mg"
    -> plural was used for "EcoScore - threatened species - Ingredients missing"
    hence, "_**en:serving-size-is-missing-digits**_" seems to be a good choice

### CODE REVIEW

File: lib/ProductOpener/DataQualityFood.pm [hunk 1]
```diff
@@ -659,6 +659,9 @@ sub check_nutrition_data ($product_ref) {
 					"en:nutrition-data-prepared-without-category-dried-products-to-be-rehydrated";
 			}
 		}
+		if ((defined $product_ref->{serving_size}) && (($product_ref->{serving_size} =~ /(\d+)/) eq "")) {
```
Comment (stephanegigandet) [inline lib/ProductOpener/DataQualityFood.pm:662]: The test is better written this way I think:
- use !~ //   -> indicates it does not match
- check that the field is not an empty string
- check if there's one digit /\d/

```suggestion
		if ((defined $product_ref->{serving_size}) and ($product_ref->{serving_size} ne "") and ($product_ref->{serving_size} !~ /\d/) ) {
```

**Comment by benbenben2** [comment]
Thank you @stephanegigandet
Your suggested code is working fine (tested with gitpod).
But for perltidy, no idea why it is failing (tried changing formatting, it is working with only 2 conditions but not with 3) :-(

**Comment by stephanegigandet** [comment]
> But for perltidy, no idea why it is failing (tried changing formatting, it is working with only 2 conditions but not with 3) :-(

Hi @benbenben2, you need to run "make lint", and perltidy will change the indentation, split long lines etc.

**Review by stephanegigandet — APPROVED** [review]
Thank you!

**Comment by CharlesNepote** [comment]
Didn't have time to thank you, it's great!

**Pull request merged**